File: src/translator/zh2en/provider/sting.ts
import { DATA as POE_DATA } from "../../../data/poe/index.js";
import type { ClientString } from "../../../data/poe/types.js";

export class StringProvider {
    private readonly idx = new Map<string, ClientString>();

    constructor() {
        for (const str of POE_DATA.strings) {
            this.idx.set(str.id, str);
        }
    }

    provide(id: string): ClientString | undefined {
        return this.idx.get(id);
    }

    mustProvide(id: string): ClientString {
        const str = this.provide(id);
        if (!str) {
            throw new Error(`id not found: ${id}`);
        }
        return str;
    }
}

File: go/translator/zh2en/provider/sting.go
package provider

import (
	"github.com/cn-poe-community/cn-poe-utils/go/data/poe"
)

type StringProvider struct {
	idx map[string]*poe.ClientString
}

func NewStringProvider() *StringProvider {
	idx := make(map[string]*poe.ClientString)
	for i := range poe.DATA.Strings {
		item := &poe.DATA.Strings[i]
		idx[item.Id] = item
	}
	return &StringProvider{
		idx: idx,
	}
}

func (p *StringProvider) MustProvide(id string) *poe.ClientString {
	if item, ok := p.idx[id]; ok {
		return item
	}
	panic("string not found of id: " + id)
}
